### ISSUE TITLE
Wrong type inferred for indirect call of generic function 

### ISSUE THREAD

**Issue by tp** [Microsoft/TypeScript#29638]
**TypeScript Version:** 3.2.2

**Search Terms:** Generics, lookahead, type inference

**Code**

```ts
function call<P1, R>(f: (p1: P1) => R, p1: P1): R {
    return f(p1);
}

function identity<T>(o: T): T {
    return o;
}

const x = call(identity, 1);


// When the parameters are reversed, it works:

function callReversed<P1, R>(p1: P1, f: (p1: P1) => R): R {
    return f(p1);
}

const y = callReversed(1, identity);
```

**Expected behavior:**

`x` is of type `number`

**Actual behavior:**

`x` is of type `{}`
only `y` is of type number

**Playground Link:** [Playground link](https://www.typescriptlang.org/play/#src=function%20call%3CP1%2C%20R%3E(f%3A%20(p1%3A%20P1)%20%3D%3E%20R%2C%20p1%3A%20P1)%3A%20R%20%7B%0A%20%20%20%20return%20f(p1)%3B%0A%7D%0A%0Afunction%20identity%3CT%3E(o%3A%20T)%3A%20T%20%7B%0A%20%20%20%20return%20o%3B%0A%7D%0A%0Aconst%20x%20%3D%20call(identity%2C%201)%3B%0A%0A%0A%2F%2F%20Reversed%3A%0A%0Afunction%20callReversed%3CP1%2C%20R%3E(p1%3A%20P1%2C%20f%3A%20(p1%3A%20P1)%20%3D%3E%20R)%3A%20R%20%7B%0A%20%20%20%20return%20f(p1)%3B%0A%7D%0A%0Aconst%20y%20%3D%20callReversed(1%2C%20identity)%3B%0A)

**Related Issues:** I didn't find anything related, but maybe I have been using the wrong search term ("lookahead").

**Comment by tp** [comment]
Would this also be fixed by #30215?

**Comment by tp** [comment]
@RyanCavanaugh Why did you close this if the reported issue still persists?

**Comment by RyanCavanaugh** [comment]
@tp Issues are closed if there isn't an identifiable next set of work to do. In this case, we believe the only next step would be a near-complete rewrite of generic inference, which is already tracked at #30134.

**Comment by ahejlsberg** [comment]
Actually, this issue _is_ fixed with the latest commits in #30215.
